.circleci/config.yml

```diff
@@ -0,0 +1,171 @@
+              then
+              ${REPO_NAME}="https://www.github.com/${CONTAINER_NAME}"
+            fi
+            echo "Container name is ${CONTAINER_NAME}"
+            echo "Repository name is ${REPO_NAME}"
+            echo "2. Running jupyter-repo2docker..."
+            echo "jupyter-repo2docker --debug --user-name username --user-id 1000 --no-run --image-name ${CONTAINER_NAME} ${REPO_NAME}"
+            jupyter-repo2docker --debug --user-name username --user-id 1000 --no-run --image-name "${CONTAINER_NAME}" "${REPO_NAME}"
+            echo "3. Running without server (example)"
+            docker run -td --name repo2docker --entrypoint /bin/bash "${CONTAINER_NAME}"
+            docker ps
+            docker images
+      - run:
+          name: Docker Save
+          no_output_timeout: 40m
+          command: |
+            # If not set, define CONTAINER_NAME
+            if [ ! -n "${CONTAINER_NAME:-}" ]
+                then
+                    CONTAINER_NAME="${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}"
+            fi
+            echo "Saving ${CONTAINER_NAME} to container.tar.gz"
+            mkdir -p /tmp/cache
+            docker save ${CONTAINER_NAME}:latest \
+              | pigz -2 -p 3 > /tmp/cache/container.tar.gz
+      - persist_to_workspace:
+          root: /tmp
+          paths:
+            - cache/container.tar.gz
+            - src
+
+  update_cache:
+    machine:
+      # Ubuntu 14.04 with Docker 17.10.0-ce
+      image: circleci/classic:201711-01
+    working_directory: /tmp/src
+    steps:
+      - attach_workspace:
+          at: /tmp
+      - save_cache:
+         key: docker-v1-{{ .Branch }}
+         paths:
+            - /tmp/cache/container.tar.gz
+
+  deploy:
+    machine:
+      image: circleci/classic:201711-01
+    working_directory: /tmp/src
+    steps:
+      - attach_workspace:
+          at: /tmp
+      - run:
+          name: Load Docker image layer cache
+          no_output_timeout: 30m
+          command: |
+            echo "Working directory is ${PWD}"
+            docker info
+            set +o pipefail
+            if [ -f /tmp/cache/container.tar.gz ]; then
+              sudo apt update && sudo apt -y install pigz
+              pigz -d --stdout /tmp/cache/container.tar.gz | docker load
+              docker images
+            fi
+      - run:
+          name: Deploy to Docker Hub
+          no_output_timeout: 40m
+          command: |
+            echo "Looking for Docker deployment options"
+            # If not set, define CONTAINER_NAME
+            if [[ ! -n "${CONTAINER_NAME:-}" ]]
+                then
+                    CONTAINER_NAME="${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}"
+            fi
+            echo "Container name set to ${CONTAINER_NAME}"
+            if [[ -n "$DOCKER_PASS" ]]; then
+                  docker login -u $DOCKER_USER -p $DOCKER_PASS
+                  docker push ${CONTAINER_NAME}:latest
```

Comment:
I'd push using the first seven (or so) digits of the commit hash as tag instead of `latest`. Get people in the habit of explicitly specifying version :)

Reply:
We should document somewhere (even if the `.rst` docs already do so) here or when CONTAINER_NAME is defined that you have to own the docker hub organisation that matches your GitHub organisation for this setup to work. At least I think that is true.

Reply:
Haha, sure thing! How about if the user explicitly defines a `DOCKER_TAG` we use that, but if not, add the commit tag instead? It looks like I can grab the commit with `CIRCLE_SHA1`

Reply:
I'd use the SHA1 as default. If people hate it they can redefine that bit of circleci config. I think that dockerhub always has a `latest` that points to the latest image.

Reply:
okay just pushed a change for `DOCKER_TAG`, let's see if it works!